### PR TITLE
M2P-258 Add support for product add-ons

### PR DESCRIPTION
**Pull request by dumega**
# Description
Add support for product add-ons

Fixes: https://boltpay.atlassian.net/browse/M2P-258

#changelog Add support for product add-ons

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Review by danac-gs — APPROVED** [review]
LGTM

File: Model/Api/UpdateCartItemTrait.php [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                sprintf('The price of item [%s] does not match product price.', $itemToAdd['product_id']),
+                422
+            );
+
+            return false;
+        }
+        
+        if (!$this->stockState->verifyStock($itemToAdd['product_id'])) {
+            $this->sendErrorResponse(
+                BoltErrorResponse::ERR_PPC_OUT_OF_STOCK,
```
Comment (vitaliyreznikov) [inline Model/Api/UpdateCartItemTrait.php:121]: All error codes should start from 66

**Review by vitaliyreznikov — COMMENTED** [review]
Could you please fix the test?

File: Helper/Cart.php [hunk 2]
```diff
@@ -1354,6 +1354,7 @@ function ($item) use ($imageHelper, &$totalAmount, &$diff, $storeId, $currencyCo
                 $product['unit_price']   = CurrencyUtils::toMinor($unitPrice, $currencyCode);
                 $product['quantity']     = round($item->getQty());
                 $product['sku']          = trim($item->getSku());
+                $product['quote_item_id']= $item->getId();
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1357]: I am not sure it's the best way. We need `quote_item_id` only when we remove the item from the cart, but this line includes it to bolt cart as well.
As I see right now we use only reference, quantity, and quote_item_id. I suggest creating another helper that returns only the information we needed.
It's not only the way, we can use some other approaches as well, but lets save bolt cart unaltered.

File: Model/Api/UpdateCartItemTrait.php (hunk 1 above)
Comment (dumega) [reply Model/Api/UpdateCartItemTrait.php]: fixed

**Review by vitaliyreznikov — CHANGES_REQUESTED** [review]

File: Helper/Cart.php (hunk 2 above)
Comment (dumega) [reply Helper/Cart.php]: It is too specific, so I create a method in UpdateCartCommon class, please review again

File: Model/Api/UpdateCartCommon.php [hunk 3]
```diff
@@ -281,7 +287,44 @@ protected function getRequestContent()
...
+
+        $products = array_map(
+            function ($item) {
+                $product = [];
+
+                $product['reference']    = $item->getProductId();
+                $product['quantity']     = round($item->getQty());
+                $product['quote_item_id']= $item->getId();
+
+                return  $product;
```
Comment (vitaliyreznikov) [inline Model/Api/UpdateCartCommon.php:321]: two spaces

**Review by vitaliyreznikov — APPROVED** [review]

File: Model/Api/UpdateCartCommon.php (hunk 3 above)
Comment (dumega) [reply Model/Api/UpdateCartCommon.php]: fixed

**Review by vitaliyreznikov — APPROVED** [review]

**Review by danac-gs — APPROVED** [review]

**Pull request merged**